doc/specs/#4999 - Improved keyboard handling in Conpty.md

```diff
@@ -0,0 +1,452 @@
+  or released (though, less than a full `INPUT_RECORD` worth of data). Instead
+  of input being sent to the client as a CSI or SS3 sequence, this `kitty` mode
+  uses "Application Program-Command" (or "APC) sequences , prefixed with `^[_`.
+* [iTerm2](https://www.iterm2.com/documentation-escape-codes.html) has a region
+  of OSC's that they've carved for themselves all starting with the same initial
+  parameter, `1337`. They then have a number of commands that all use the second
+  parameter to indicate what command specific to iTerm2 they're actually
+  implementing.
+
+### Requesting `win32-input-mode`
+
+An application can request `win32-input-mode` with the following OSC sequence
+
+```
+  ^[ ] 1000 ; 1 ; Ps ST
+
+         Ps: Whether to enable or disable win32-input-mode. If omitted, defaults to '0'.
+             0: Disable win32-input-mode
+             1: Enable win32-input-mode
+```
+
+OSC 1000 seemed to be unused according to [this discussion of used
+OSCs](https://gitlab.freedesktop.org/terminal-wg/specifications/-/issues/10).
+This seems like a reasonable place to stake a claim for sequences used by the
+the Windows Terminal - future WT-specific sequences could also be placed as `OSC
+1000;2`, `OSC 1000;3`, etc. This pattern is also similar to the way iTerm2
+defines their own sequences.
+
+When a terminal receives a `^[]1000;1;1ST` sequence, they should switch into
+`win32-input-mode`. In `win32-input-mode`, the terminal will send keyboard input
+to the connected client application in the following format:
+
+### `win32-input-mode` sequences
+
+The `KEY_EVENT_RECORD` portion of an input record (the part that's important for
+us to encode in this feature) is defined as the following:
+
+```c++
+typedef struct _KEY_EVENT_RECORD {
+  BOOL  bKeyDown;
+  WORD  wRepeatCount;
+  WORD  wVirtualKeyCode;
+  WORD  wVirtualScanCode;
+  union {
+    WCHAR UnicodeChar;
+    CHAR  AsciiChar;
+  } uChar;
+  DWORD dwControlKeyState;
+} KEY_EVENT_RECORD;
+```
+
+To encode all of this information, I propose the following sequence. This is a
+CSI sequence with a final terminator character of `_`. This character appears
+unused as a terminator by sequences _output_ by a client application, and
+doesn't seem to be used as an _input_ sequence terminator either.
+
+```
+  ^[ [ Kd ; Rc ; Vk ; Sc ; Uc ; Cs _
+
+       Kd: the value of bKeyDown - either a '0' or '1'. If omitted, defaults to '0'.
+
+       Rc: the value of wRepeatCount - any number. If omitted, defaults to '0'.
+
+       Vk: the value of wVirtualKeyCode - any number. If omitted, defaults to '0'.
+
+       Sc: the value of wVirtualScanCode - any number. If omitted, defaults to '0'.
+
+       Uc: the decimal value of UnicodeChar - for example, NUL is "0", LF is
+           "10", the character 'A' is "65". If omitted, defaults to '0'.
+
+       Cs: the value of dwControlKeyState - any number. If omitted, defaults to '0'.
+```

```

Comment:
You may want to consider reordering some of these parameters based on how likely they are to be used. For example, if 99% of the time the repeat count is not needed, then it would make sense to put it last, and then leave it off when not required. Similarly, the control key state is probably going to be 0 a lot of the time too, so that should perhaps be second last. Even keydown will be 0 at least half the time, so that could also potentially be omitted some of the time.

Reply:
Oh jeez, I forgot about that. That's a really good point. Is it terribly inconsistent to assume that if the repeat count was omitted, it should be `1` instead of `0`? Typically, it looks like key events have `1` there. I'd love to be able to omit it if I can assume missing is 1.

Then, I'd probably make the order `^[ [ Vk ; Sc ; Uc ; Kd ; Cs ; Rc _`. So something like just an 'a' could be 

```
^[[65;30;97;1_
^[[65;30;97_
```

If I can't assume a missing repeat count is 1, then I might as well just stick it at the beginning.

Reply:
I think it's perfectly reasonable to make the default value 1 for the repeat count. Although that's assuming it can never be zero, because it's usually the case that a 0 parameter value would also map to the default value.

Reply:
So, I'm not sure what exactly happens when a repeat count of 0 is passed in. I think internally conhost says "sure buddy, 0, I'm sure what you actually meant was _1_". In my prototype, some keys like Ctrl were still passing a repeat count of 0, but it seemed to work just fine. 

I think it's _probably_ safe though to assume that 0 isn't a valid value and default to 1. Playing with my own console keyboard logger, everything I entered just had a repeat count of 1. If you really want 0, then _don't pass the event at all_...